tsl/src/bgw_policy/process_hyper_inval_api.c

```diff
@@ -0,0 +1,212 @@
+{
+	if (PG_ARGISNULL(0))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED), errmsg("config must not be NULL")));
+	}
+
+	policy_process_hyper_inval_read_and_validate_config(PG_GETARG_JSONB_P(0), NULL);
+
+	PG_RETURN_VOID();
+}
+
+static int32
+policy_process_hyper_inval_add_internal(Hypertable *ht, bool if_not_exists,
+										Interval *schedule_interval, Oid owner_id,
+										TimestampTz initial_start, const bool fixed_schedule,
+										char *timezone)
+{
+	NameData application_name, proc_name, proc_schema, check_schema, check_name, owner;
+
+	List *jobs = ts_bgw_job_find_by_proc_and_hypertable_id(POLICY_PROCESS_HYPER_INVAL_PROC_NAME,
+														   FUNCTIONS_SCHEMA_NAME,
+														   ht->fd.id);
+	if (jobs != NIL)
+	{
+		Assert(list_length(jobs) == 1);
+		if (!if_not_exists)
+			ereport(ERROR,
+					errcode(ERRCODE_DUPLICATE_OBJECT),
+					errmsg("move hypertable invalidations policy already exists for \"%s\"",
+						   get_rel_name(ht->main_table_relid)));
+		else
+			ereport(NOTICE,
+					errmsg("move hypertable invalidations policy already exists for \"%s\", "
+						   "skipping",
+						   get_rel_name(ht->main_table_relid)));
+		return -1;
+	}
+
+	namestrcpy(&application_name, "Move Hypertables Invalidation Policy");
+	namestrcpy(&proc_name, POLICY_PROCESS_HYPER_INVAL_PROC_NAME);
+	namestrcpy(&proc_schema, FUNCTIONS_SCHEMA_NAME);
+	namestrcpy(&check_name, POLICY_PROCESS_HYPER_INVAL_CHECK_NAME);
+	namestrcpy(&check_schema, FUNCTIONS_SCHEMA_NAME);
+	namestrcpy(&owner, GetUserNameFromId(owner_id, false));
+
+	JsonbParseState *parse_state = NULL;
+
+	pushJsonbValue(&parse_state, WJB_BEGIN_OBJECT, NULL);
+	ts_jsonb_add_int32(parse_state, POLICY_CONFIG_KEY_HYPERTABLE_ID, ht->fd.id);
+
+	JsonbValue *result = pushJsonbValue(&parse_state, WJB_END_OBJECT, NULL);
+	Jsonb *config = JsonbValueToJsonb(result);
+
+	return ts_bgw_job_insert_relation(&application_name,
+									  schedule_interval,
+									  DEFAULT_MAX_RUNTIME,
+									  JOB_RETRY_UNLIMITED,
+									  schedule_interval,
+									  &proc_schema,
+									  &proc_name,
+									  &check_schema,
+									  &check_name,
+									  owner_id,
+									  true,
+									  fixed_schedule,
+									  ht->fd.id,
+									  config,
+									  initial_start,
+									  timezone);
+}
+
+Datum
+policy_process_hyper_inval_add(PG_FUNCTION_ARGS)
+{
+	Oid ht_oid = PG_GETARG_OID(0);
+	Interval *schedule_interval = PG_GETARG_INTERVAL_P(1);
+	const bool if_not_exists = PG_GETARG_BOOL(2);
+	TimestampTz initial_start = PG_ARGISNULL(3) ? DT_NOBEGIN : PG_GETARG_TIMESTAMPTZ(3);
+	const bool fixed_schedule = !PG_ARGISNULL(3);
+	char *valid_timezone = NULL;
+
+	ts_feature_flag_check(FEATURE_POLICY);
+	TS_PREVENT_FUNC_IF_READ_ONLY();
+
+	if (fixed_schedule)
+	{
+		ts_bgw_job_validate_schedule_interval(schedule_interval);
+		if (TIMESTAMP_NOT_FINITE(initial_start))
+			initial_start = ts_timer_get_current_timestamp();
```

Comment:
Not a blocker but it would be nice to have a test case for this initial_start time.

Reply:
I've added an initial start that should trigger this, but for some reason it doesn't.